File: bin/spice/cmd/cluster.go
/*
Copyright 2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package cmd

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/pem"
	"fmt"
	"log/slog"
	"math/big"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/logrusorgru/aurora"
	"github.com/spf13/cobra"
	"github.com/spiceai/spiceai/bin/spice/pkg/constants"
)

const (
	caValidityYears     = 10
	clientValidityYears = 1
	caCertFilename      = "ca.crt"
	caKeyFilename       = "ca.key"
	pkiDirName          = "pki"
	caCN                = "Spice.ai CLI Root CA - DO NOT USE IN PRODUCTION"
	defaultOU           = "unknown"
)

var clusterCmd = &cobra.Command{
	Use:   "cluster",
	Short: "Cluster operations for Spice runtime",
}

var tlsCmd = &cobra.Command{
	Use:   "tls",
	Short: "TLS certificate operations for clustered mode",
}

var tlsInitCmd = &cobra.Command{
	Use:   "init",
	Short: "Initialize a test PKI infrastructure for clustered mode",
	Long: `Initialize a test PKI infrastructure by generating a new Certificate Authority (CA).

This command creates:
- A CA certificate (~/.spice/pki/ca.crt)
- A CA private key (~/.spice/pki/ca.key)

The CA certificate is valid for 10 years and uses ECDSA P-256.

WARNING: This CA is for development and testing purposes only.
         DO NOT use these certificates in production environments.`,
	Example: `
spice cluster tls init
`,
	Run: func(cmd *cobra.Command, args []string) {
		pkiDir, err := getPKIDir()
		if err != nil {
			slog.Error("failed to determine PKI directory", "error", err)
			os.Exit(1)
		}

		caCertPath := filepath.Join(pkiDir, caCertFilename)
		caKeyPath := filepath.Join(pkiDir, caKeyFilename)

		// Check if CA already exists
		if _, err := os.Stat(caCertPath); err == nil {
			cmd.Print("CA certificate already exists. Overwrite (y/n)? ")
			var confirm string
			_, _ = fmt.Scanf("%s", &confirm)
			if strings.ToLower(strings.TrimSpace(confirm)) != "y" {
				cmd.Println("Aborted.")
				return
			}
		}

		// Create PKI directory if it doesn't exist
		if err := os.MkdirAll(pkiDir, 0700); err != nil {
			slog.Error("failed to create PKI directory", "error", err)
			os.Exit(1)
		}

		ou := getOrganizationalUnit()

		// Generate CA private key
		caKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
		if err != nil {
			slog.Error("failed to generate CA private key", "error", err)
			os.Exit(1)
		}

		// Create CA certificate template
		serialNumber, err := generateSerialNumber()
		if err != nil {
			slog.Error("failed to generate serial number", "error", err)
			os.Exit(1)
		}

		notBefore := time.Now()
		notAfter := notBefore.AddDate(caValidityYears, 0, 0)

		caTemplate := x509.Certificate{
			SerialNumber: serialNumber,
			Subject: pkix.Name{
				CommonName:         caCN,
				OrganizationalUnit: []string{ou},
			},
			NotBefore:             notBefore,
			NotAfter:              notAfter,
			KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign | x509.KeyUsageDigitalSignature,
			IsCA:                  true,
			BasicConstraintsValid: true,
			MaxPathLen:            0,
			MaxPathLenZero:        true,
		}

		// Create self-signed CA certificate
		caCertDER, err := x509.CreateCertificate(rand.Reader, &caTemplate, &caTemplate, &caKey.PublicKey, caKey)
		if err != nil {
			slog.Error("failed to create CA certificate", "error", err)
			os.Exit(1)
		}

		// Write CA certificate
		if err := writePEMFile(caCertPath, "CERTIFICATE", caCertDER); err != nil {
			slog.Error("failed to write CA certificate", "error", err)
			os.Exit(1)
		}

		// Write CA private key
		caKeyDER, err := x509.MarshalECPrivateKey(caKey)
		if err != nil {
			slog.Error("failed to marshal CA private key", "error", err)
			os.Exit(1)
		}
		if err := writePEMFile(caKeyPath, "EC PRIVATE KEY", caKeyDER); err != nil {
			slog.Error("failed to write CA private key", "error", err)
			os.Exit(1)
		}

		cmd.Println()
		cmd.Println(fmt.Sprintf("%s", aurora.BrightGreen("Test PKI infrastructure initialized successfully!")))
		cmd.Println()
		cmd.Println(fmt.Sprintf("CA Certificate: %s", caCertPath))
		cmd.Println(fmt.Sprintf("CA Private Key: %s", caKeyPath))
		cmd.Println(fmt.Sprintf("Validity:       %d years (until %s)", caValidityYears, notAfter.Format("2006-01-02")))
		cmd.Println(fmt.Sprintf("OU:             %s", ou))
		cmd.Println()
		cmd.Println(fmt.Sprintf("%s", aurora.BrightYellow("⚠️  WARNING: This CA is for development and testing only.")))
		cmd.Println(fmt.Sprintf("%s", aurora.BrightYellow("            DO NOT use these certificates in production!")))
		cmd.Println()
		cmd.Println("Next steps:")
		cmd.Println(fmt.Sprintf("  Run %s to create a certificate for a cluster member.", aurora.BrightCyan("spice cluster tls add <client-name>")))
	},
}

var tlsAddCmd = &cobra.Command{
	Use:   "add <client-name>",
	Short: "Create a new client certificate signed by the CA",
	Long: `Create a new client certificate and private key signed by the CA.

This command creates:
- A client certificate (~/.spice/pki/<client-name>.crt)
- A client private key (~/.spice/pki/<client-name>.key)

The client certificate is valid for 1 year and uses ECDSA P-256.
The certificate includes localhost and 127.0.0.1 as Subject Alternative Names (SANs).

The CA must be initialized first using 'spice cluster tls init'.`,
	Example: `
spice cluster tls add node1
spice cluster tls add my-spice-instance
spice cluster tls add node1 --host myserver.example.com
`,
	Args: cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		clientName := args[0]

		if clientName == "" {
			slog.Error("client name cannot be empty")
			os.Exit(1)
		}

		// Validate client name (alphanumeric, hyphens, underscores)
		for _, c := range clientName {
			if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && (c < '0' || c > '9') && c != '-' && c != '_' {
				slog.Error("client name can only contain letters, numbers, hyphens, and underscores")
				os.Exit(1)
			}
		}

		pkiDir, err := getPKIDir()
		if err != nil {
			slog.Error("failed to determine PKI directory", "error", err)
			os.Exit(1)
		}

		caCertPath := filepath.Join(pkiDir, caCertFilename)
		caKeyPath := filepath.Join(pkiDir, caKeyFilename)
		clientCertPath := filepath.Join(pkiDir, clientName+".crt")
		clientKeyPath := filepath.Join(pkiDir, clientName+".key")

		// Check if CA exists
		if _, err := os.Stat(caCertPath); os.IsNotExist(err) {
			slog.Error("CA certificate not found. Run 'spice cluster tls init' first.")
			os.Exit(1)
		}
		if _, err := os.Stat(caKeyPath); os.IsNotExist(err) {
			slog.Error("CA private key not found. Run 'spice cluster tls init' first.")
			os.Exit(1)
		}

		// Check if client cert already exists
		if _, err := os.Stat(clientCertPath); err == nil {
			cmd.Printf("Certificate for '%s' already exists. Overwrite (y/n)? ", clientName)
			var confirm string
			_, _ = fmt.Scanf("%s", &confirm)
			if strings.ToLower(strings.TrimSpace(confirm)) != "y" {
				cmd.Println("Aborted.")
				return
			}
		}

		// Load CA certificate
		caCertPEM, err := os.ReadFile(caCertPath)
		if err != nil {
			slog.Error("failed to read CA certificate", "error", err)
			os.Exit(1)
		}
		caCertBlock, _ := pem.Decode(caCertPEM)
		if caCertBlock == nil {
			slog.Error("failed to decode CA certificate PEM")
			os.Exit(1)
		}
		caCert, err := x509.ParseCertificate(caCertBlock.Bytes)
		if err != nil {
			slog.Error("failed to parse CA certificate", "error", err)
			os.Exit(1)
		}

		// Load CA private key
		caKeyPEM, err := os.ReadFile(caKeyPath)
		if err != nil {
			slog.Error("failed to read CA private key", "error", err)
			os.Exit(1)
		}
		caKeyBlock, _ := pem.Decode(caKeyPEM)
		if caKeyBlock == nil {
			slog.Error("failed to decode CA private key PEM")
			os.Exit(1)
		}
		caKey, err := x509.ParseECPrivateKey(caKeyBlock.Bytes)
		if err != nil {
			slog.Error("failed to parse CA private key", "error", err)
			os.Exit(1)
		}

		// Generate client private key
		clientKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
		if err != nil {
			slog.Error("failed to generate client private key", "error", err)
			os.Exit(1)
		}

		// Create client certificate template
		serialNumber, err := generateSerialNumber()
		if err != nil {
			slog.Error("failed to generate serial number", "error", err)
			os.Exit(1)
		}

		notBefore := time.Now()
		notAfter := notBefore.AddDate(clientValidityYears, 0, 0)

		// Build SANs - always include localhost and 127.0.0.1
		dnsNames := []string{"localhost"}
		ipAddresses := []net.IP{net.ParseIP("127.0.0.1")}

		// Add optional host to SANs
		hostFlag, _ := cmd.Flags().GetString("host")
		if hostFlag != "" {
			// Check if it's an IP address or DNS name
			if ip := net.ParseIP(hostFlag); ip != nil {
				ipAddresses = append(ipAddresses, ip)
			} else {
				dnsNames = append(dnsNames, hostFlag)
			}
		}

		clientTemplate := x509.Certificate{
			SerialNumber: serialNumber,
			Subject: pkix.Name{
				CommonName: clientName,
			},
			NotBefore:             notBefore,
			NotAfter:              notAfter,
			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
			BasicConstraintsValid: true,
			IsCA:                  false,
			DNSNames:              dnsNames,
			IPAddresses:           ipAddresses,
		}

		// Create client certificate signed by CA
		clientCertDER, err := x509.CreateCertificate(rand.Reader, &clientTemplate, caCert, &clientKey.PublicKey, caKey)
		if err != nil {
			slog.Error("failed to create client certificate", "error", err)
			os.Exit(1)
		}

		// Write client certificate
		if err := writePEMFile(clientCertPath, "CERTIFICATE", clientCertDER); err != nil {
			slog.Error("failed to write client certificate", "error", err)
			os.Exit(1)
		}

		// Write client private key
		clientKeyDER, err := x509.MarshalECPrivateKey(clientKey)
		if err != nil {
			slog.Error("failed to marshal client private key", "error", err)
			os.Exit(1)
		}
		if err := writePEMFile(clientKeyPath, "EC PRIVATE KEY", clientKeyDER); err != nil {
			slog.Error("failed to write client private key", "error", err)
			os.Exit(1)
		}

		cmd.Println()
		cmd.Println(fmt.Sprintf("%s", aurora.BrightGreen(fmt.Sprintf("Certificate for '%s' created successfully!", clientName))))
		cmd.Println()
		cmd.Println(fmt.Sprintf("Certificate: %s", clientCertPath))
		cmd.Println(fmt.Sprintf("Private Key: %s", clientKeyPath))
		cmd.Println(fmt.Sprintf("Validity:    %d year (until %s)", clientValidityYears, notAfter.Format("2006-01-02")))
		cmd.Println(fmt.Sprintf("CN:          %s", clientName))
		cmd.Println(fmt.Sprintf("DNS SANs:    %s", strings.Join(dnsNames, ", ")))
		ipStrs := make([]string, len(ipAddresses))
		for i, ip := range ipAddresses {
			ipStrs[i] = ip.String()
		}
		cmd.Println(fmt.Sprintf("IP SANs:     %s", strings.Join(ipStrs, ", ")))
	},
}

// getPKIDir returns the path to the PKI directory (~/.spice/pki)
func getPKIDir() (string, error) {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("failed to get home directory: %w", err)
	}
	return filepath.Join(homeDir, constants.DotSpice, pkiDirName), nil
}

// getOrganizationalUnit attempts to get the user's email from git config,
// falling back to "unknown" if not available
func getOrganizationalUnit() string {
	out, err := exec.Command("git", "config", "--get", "user.email").Output()
	if err != nil {
		return defaultOU
	}
	email := strings.TrimSpace(string(out))
	if email == "" {
		return defaultOU
	}
	return email
}

// generateSerialNumber generates a random serial number for certificates
func generateSerialNumber() (*big.Int, error) {
	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
	return rand.Int(rand.Reader, serialNumberLimit)
}

// writePEMFile writes data as a PEM-encoded file with secure permissions
func writePEMFile(path string, pemType string, data []byte) error {
	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			slog.Error("failed to close file", "file", path, "error", err)
		}
	}()

	block := &pem.Block{
		Type:  pemType,
		Bytes: data,
	}
	if err := pem.Encode(file, block); err != nil {
		return fmt.Errorf("failed to encode PEM: %w", err)
	}

	// Ensure secure permissions
	if err := file.Chmod(0600); err != nil {
		return fmt.Errorf("failed to set file permissions: %w", err)
	}

	return nil
}

func init() {
	tlsAddCmd.Flags().String("host", "", "Host to include in Subject Alternative Names (e.g., myserver.example.com)")
	tlsCmd.AddCommand(tlsInitCmd)
	tlsCmd.AddCommand(tlsAddCmd)
	clusterCmd.AddCommand(tlsCmd)
	RootCmd.AddCommand(clusterCmd)
}

File: bin/spice/src/commands/cluster.rs
/*
Copyright 2024-2026 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! `spice cluster` command - Cluster operations for Spice runtime.

use crate::error::{
    ConfigIoSnafu, CreateDirectorySnafu, HomeDirectoryNotFoundSnafu, InvalidArgumentSnafu, Result,
};
use ansi_colors::Color;
use clap::{Args, Subcommand};
use rcgen::{
    BasicConstraints, CertificateParams, DnType, ExtendedKeyUsagePurpose, IsCa, KeyPair,
    KeyUsagePurpose, SanType,
};
use snafu::{ResultExt, ensure};
use std::fs;
use std::io::{self, Write};
use std::net::IpAddr;
use std::path::PathBuf;
use std::process::Command;
use time::{Duration, OffsetDateTime};

const CA_VALIDITY_YEARS: i64 = 10;
const CLIENT_VALIDITY_YEARS: i64 = 1;
const CA_CERT_FILENAME: &str = "ca.crt";
const CA_KEY_FILENAME: &str = "ca.key";
const PKI_DIR_NAME: &str = "pki";
const DOT_SPICE: &str = ".spice";
const CA_CN: &str = "Spice.ai CLI Root CA - DO NOT USE IN PRODUCTION";
const DEFAULT_OU: &str = "unknown";

/// Arguments for the `cluster` command.
#[derive(Args, Debug)]
pub struct ClusterArgs {
    #[command(subcommand)]
    pub command: ClusterCommands,
}

/// Cluster subcommands.
#[derive(Subcommand, Debug)]
pub enum ClusterCommands {
    /// TLS certificate operations for clustered mode
    Tls(TlsArgs),
}

/// TLS subcommand arguments.
#[derive(Args, Debug)]
pub struct TlsArgs {
    #[command(subcommand)]
    pub command: TlsCommands,
}

/// TLS subcommands.
#[derive(Subcommand, Debug)]
pub enum TlsCommands {
    /// Initialize a test PKI infrastructure for clustered mode
    #[command(
        long_about = "Initialize a test PKI infrastructure by generating a new Certificate Authority (CA).\n\n\
        This command creates:\n\
        - A CA certificate (~/.spice/pki/ca.crt)\n\
        - A CA private key (~/.spice/pki/ca.key)\n\n\
        The CA certificate is valid for 10 years and uses ECDSA P-256.\n\n\
        WARNING: This CA is for development and testing purposes only.\n\
                 DO NOT use these certificates in production environments."
    )]
    Init,

    /// Create a new client certificate signed by the CA
    #[command(
        long_about = "Create a new client certificate and private key signed by the CA.\n\n\
        This command creates:\n\
        - A client certificate (~/.spice/pki/<client-name>.crt)\n\
        - A client private key (~/.spice/pki/<client-name>.key)\n\n\
        The client certificate is valid for 1 year and uses ECDSA P-256.\n\
        The certificate includes localhost and 127.0.0.1 as Subject Alternative Names (SANs).\n\n\
        The CA must be initialized first using 'spice cluster tls init'."
    )]
    Add {
        /// Name of the client/node
        client_name: String,

        /// Additional host to include in Subject Alternative Names
        #[arg(long)]
        host: Option<String>,
    },
}

/// Execute the `cluster` command.
///
/// # Errors
///
/// Returns an error if certificate generation or file I/O fails.
pub fn execute(args: &ClusterArgs) -> Result<()> {
    match &args.command {
        ClusterCommands::Tls(tls_args) => match &tls_args.command {
            TlsCommands::Init => execute_tls_init(),
            TlsCommands::Add { client_name, host } => execute_tls_add(client_name, host.as_deref()),
        },
    }
}

/// Get the PKI directory (~/.spice/pki).
fn get_pki_dir() -> Result<PathBuf> {
    let home_dir = dirs::home_dir().ok_or_else(|| HomeDirectoryNotFoundSnafu.build())?;
    Ok(home_dir.join(DOT_SPICE).join(PKI_DIR_NAME))
}

/// Get the organizational unit from git config email, or default.
fn get_organizational_unit() -> String {
    Command::new("git")
        .args(["config", "--get", "user.email"])
        .output()
        .ok()
        .and_then(|output| {
            if output.status.success() {
                String::from_utf8(output.stdout)
                    .ok()
                    .map(|s| s.trim().to_string())
                    .filter(|s| !s.is_empty())
            } else {
                None
            }
        })
        .unwrap_or_else(|| DEFAULT_OU.to_string())
}

/// Prompt user for confirmation.
fn confirm_overwrite(what: &str) -> bool {
    print!("{what} already exists. Overwrite (y/n)? ");
    let _ = io::stdout().flush();

    let mut input = String::new();
    if io::stdin().read_line(&mut input).is_ok() {
        input.trim().eq_ignore_ascii_case("y")
    } else {
        false
    }
}

/// Validate client name (alphanumeric, hyphens, underscores).
fn validate_client_name(name: &str) -> Result<()> {
    ensure!(
        !name.is_empty(),
        InvalidArgumentSnafu {
            message: "client name cannot be empty"
        }
    );

    ensure!(
        name.chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_'),
        InvalidArgumentSnafu {
            message: "client name can only contain letters, numbers, hyphens, and underscores"
        }
    );

    Ok(())
}

/// Initialize the PKI infrastructure with a new CA.
fn execute_tls_init() -> Result<()> {
    let pki_dir = get_pki_dir()?;
    let ca_cert_path = pki_dir.join(CA_CERT_FILENAME);
    let ca_key_path = pki_dir.join(CA_KEY_FILENAME);

    // Check if CA already exists
    if ca_cert_path.exists() && !confirm_overwrite("CA certificate") {
        println!("Aborted.");
        return Ok(());
    }

    // Create PKI directory
    fs::create_dir_all(&pki_dir).context(CreateDirectorySnafu {
        path: pki_dir.clone(),
    })?;

    // Set secure permissions on PKI directory
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let permissions = fs::Permissions::from_mode(0o700);
        fs::set_permissions(&pki_dir, permissions)
            .context(CreateDirectorySnafu { path: pki_dir })?;
    }

    let ou = get_organizational_unit();

    // Generate CA certificate and key
    let not_before = OffsetDateTime::now_utc();
    let not_after = not_before + Duration::days(CA_VALIDITY_YEARS * 365);

    let mut ca_params = CertificateParams::default();
    ca_params.distinguished_name.push(DnType::CommonName, CA_CN);
    ca_params
        .distinguished_name
        .push(DnType::OrganizationalUnitName, &ou);
    ca_params.not_before = not_before;
    ca_params.not_after = not_after;
    ca_params.is_ca = IsCa::Ca(BasicConstraints::Constrained(0));
    ca_params.key_usages = vec![
        KeyUsagePurpose::DigitalSignature,
        KeyUsagePurpose::KeyCertSign,
        KeyUsagePurpose::CrlSign,
    ];

    let ca_key_pair = KeyPair::generate().map_err(|e| crate::error::Error::InvalidArgument {
        message: format!("Failed to generate CA key pair: {e}"),
    })?;

    let ca_cert =
        ca_params
            .self_signed(&ca_key_pair)
            .map_err(|e| crate::error::Error::InvalidArgument {
                message: format!("Failed to create CA certificate: {e}"),
            })?;

    // Write CA certificate
    fs::write(&ca_cert_path, ca_cert.pem()).context(ConfigIoSnafu {
        operation: "write",
        path: ca_cert_path.clone(),
    })?;

    // Write CA private key
    fs::write(&ca_key_path, ca_key_pair.serialize_pem()).context(ConfigIoSnafu {
        operation: "write",
        path: ca_key_path.clone(),
    })?;

    // Set secure permissions on files
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let permissions = fs::Permissions::from_mode(0o600);
        fs::set_permissions(&ca_cert_path, permissions.clone()).context(ConfigIoSnafu {
            operation: "set permissions on",
            path: ca_cert_path.clone(),
        })?;
        fs::set_permissions(&ca_key_path, permissions).context(ConfigIoSnafu {
            operation: "set permissions on",
            path: ca_key_path.clone(),
        })?;
    }

    println!();
    println!(
        "{}",
        Color::Green.paint("Test PKI infrastructure initialized successfully!")
    );
    println!();
    println!("CA Certificate: {}", ca_cert_path.display());
    println!("CA Private Key: {}", ca_key_path.display());
    println!(
        "Validity:       {CA_VALIDITY_YEARS} years (until {})",
        not_after.date()
    );
    println!("OU:             {ou}");
    println!();
    println!(
        "{}",
        Color::Yellow.paint("WARNING: This CA is for development and testing only.")
    );
    println!(
        "{}",
        Color::Yellow.paint("         DO NOT use these certificates in production!")
    );
    println!();
    println!("Next steps:");
    println!(
        "  Run {} to create a certificate for a cluster member.",
        Color::Cyan.paint("spice cluster tls add <client-name>")
    );

    Ok(())
}

/// Create a client certificate signed by the CA.
fn execute_tls_add(client_name: &str, host: Option<&str>) -> Result<()> {
    validate_client_name(client_name)?;

    let pki_dir = get_pki_dir()?;
    let ca_cert_path = pki_dir.join(CA_CERT_FILENAME);
    let ca_key_path = pki_dir.join(CA_KEY_FILENAME);
    let client_cert_path = pki_dir.join(format!("{client_name}.crt"));
    let client_key_path = pki_dir.join(format!("{client_name}.key"));

    // Check if CA exists
    ensure!(
        ca_cert_path.exists(),
        InvalidArgumentSnafu {
            message: "CA certificate not found. Run 'spice cluster tls init' first."
        }
    );
    ensure!(
        ca_key_path.exists(),
        InvalidArgumentSnafu {
            message: "CA private key not found. Run 'spice cluster tls init' first."
        }
    );

    // Check if client cert already exists
    if client_cert_path.exists() && !confirm_overwrite(&format!("Certificate for '{client_name}'"))
    {
        println!("Aborted.");
        return Ok(());
    }

    // Load CA certificate and key
    let ca_cert_pem = fs::read_to_string(&ca_cert_path).context(ConfigIoSnafu {
        operation: "read",
        path: ca_cert_path,
    })?;
    let ca_key_pem = fs::read_to_string(&ca_key_path).context(ConfigIoSnafu {
        operation: "read",
        path: ca_key_path,
    })?;

    // Parse the CA key
    let ca_key_pair =
        KeyPair::from_pem(&ca_key_pem).map_err(|e| crate::error::Error::InvalidArgument {
            message: format!("Failed to parse CA private key: {e}"),
        })?;

    // Parse the CA certificate from PEM to reconstruct it for signing
    let ca_params = CertificateParams::from_ca_cert_pem(&ca_cert_pem).map_err(|e| {
        crate::error::Error::InvalidArgument {
            message: format!("Failed to parse CA certificate: {e}"),
        }
    })?;

    let ca_cert =
        ca_params
            .self_signed(&ca_key_pair)
            .map_err(|e| crate::error::Error::InvalidArgument {
                message: format!("Failed to reconstruct CA certificate: {e}"),
            })?;

    // Generate client certificate
    let not_before = OffsetDateTime::now_utc();
    let not_after = not_before + Duration::days(CLIENT_VALIDITY_YEARS * 365);

    // Build SANs - always include localhost and 127.0.0.1
    let mut dns_names = vec!["localhost".to_string()];
    let mut ip_addresses: Vec<IpAddr> = vec![IpAddr::V4(std::net::Ipv4Addr::LOCALHOST)];

    // Add optional host to SANs
    if let Some(host) = host {
        if let Ok(ip) = host.parse::<IpAddr>() {
            ip_addresses.push(ip);
        } else {
            dns_names.push(host.to_string());
        }
    }

    let mut client_params = CertificateParams::default();
    client_params
        .distinguished_name
        .push(DnType::CommonName, client_name);
    client_params.not_before = not_before;
    client_params.not_after = not_after;
    client_params.is_ca = IsCa::NoCa;
    client_params.key_usages = vec![
        KeyUsagePurpose::DigitalSignature,
        KeyUsagePurpose::KeyEncipherment,
    ];
    client_params.extended_key_usages = vec![
        ExtendedKeyUsagePurpose::ClientAuth,
        ExtendedKeyUsagePurpose::ServerAuth,
    ];

    // Add SANs
    for dns in &dns_names {
        client_params
            .subject_alt_names
            .push(SanType::DnsName(dns.clone().try_into().map_err(|e| {
                crate::error::Error::InvalidArgument {
                    message: format!("Invalid DNS name '{dns}': {e}"),
                }
            })?));
    }
    for ip in &ip_addresses {
        client_params
            .subject_alt_names
            .push(SanType::IpAddress(*ip));
    }

    let client_key_pair =
        KeyPair::generate().map_err(|e| crate::error::Error::InvalidArgument {
            message: format!("Failed to generate client key pair: {e}"),
        })?;

    let client_cert = client_params
        .signed_by(&client_key_pair, &ca_cert, &ca_key_pair)
        .map_err(|e| crate::error::Error::InvalidArgument {
            message: format!("Failed to create client certificate: {e}"),
        })?;

    // Write client certificate
    fs::write(&client_cert_path, client_cert.pem()).context(ConfigIoSnafu {
        operation: "write",
        path: client_cert_path.clone(),
    })?;

    // Write client private key
    fs::write(&client_key_path, client_key_pair.serialize_pem()).context(ConfigIoSnafu {
        operation: "write",
        path: client_key_path.clone(),
    })?;

    // Set secure permissions on files
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let permissions = fs::Permissions::from_mode(0o600);
        fs::set_permissions(&client_cert_path, permissions.clone()).context(ConfigIoSnafu {
            operation: "set permissions on",
            path: client_cert_path.clone(),
        })?;
        fs::set_permissions(&client_key_path, permissions).context(ConfigIoSnafu {
            operation: "set permissions on",
            path: client_key_path.clone(),
        })?;
    }

    println!();
    println!(
        "{}",
        Color::Green.paint(format!(
            "Certificate for '{client_name}' created successfully!"
        ))
    );
    println!();
    println!("Certificate: {}", client_cert_path.display());
    println!("Private Key: {}", client_key_path.display());
    println!(
        "Validity:    {CLIENT_VALIDITY_YEARS} year (until {})",
        not_after.date()
    );
    println!("CN:          {client_name}");
    println!("DNS SANs:    {}", dns_names.join(", "));
    println!(
        "IP SANs:     {}",
        ip_addresses
            .iter()
            .map(ToString::to_string)
            .collect::<Vec<_>>()
            .join(", ")
    );

    Ok(())
}
